File: src/main/java/cc/moecraft/logger/TimingLogger.java
package cc.moecraft.logger;

/**
 * 此类由 Hykilpikonna 在 2018/07/10 创建!
 * Created by Hykilpikonna on 2018/07/10!
 * Github: https://github.com/hykilpikonna
 * Meow!
 *
 * @author Hykilpikonna
 */
public class TimingLogger
{
    private HyLogger logger;
    private Long startingTime;

    protected TimingLogger(HyLogger logger)
    {
        this.logger = logger;
    }

    public TimingLogger time()
    {
        logger.log(getMilliseconds() + "ms");
        return this;
    }

    public TimingLogger timeAndReset()
    {
        return time().reset();
    }

    public TimingLogger timeNano()
    {
        logger.log(getTime() + " nanos");
        return this;
    }

    public TimingLogger timeNanoAndReset()
    {
        return timeNano().reset();
    }

    public long getTime()
    {
        return System.nanoTime() - startingTime;
    }

    public double getMilliseconds()
    {
        return ((double) getTime()) / 1000000d;
    }

    public TimingLogger reset()
    {
        startingTime = System.nanoTime();
        return this;
    }

    public TimingLogger init()
    {
        return reset();
    }

    public void clear()
    {
        startingTime = null;
    }
}

File: src/main/kotlin/cc/moecraft/logger/TimingLogger.kt
package cc.moecraft.logger

/**
 * 此类由 Hykilpikonna 在 2018/07/10 创建!
 * Created by Hykilpikonna on 2018/07/10!
 * Github: https://github.com/hykilpikonna
 * QQ: admin@moecraft.cc -OR- 871674895
 *
 * @author Hykilpikonna
 */
class TimingLogger(private val logger: HyLogger)
{
    private var startingTime: Long? = null
    fun time(): TimingLogger
    {
        logger.log(milliseconds.toString() + "ms")
        return this
    }

    fun timeAndReset(): TimingLogger
    {
        return time().reset()
    }

    fun timeNano(): TimingLogger
    {
        logger.log("$time nanos")
        return this
    }

    fun timeNanoAndReset(): TimingLogger
    {
        return timeNano().reset()
    }

    val time: Long
        get() = System.nanoTime() - startingTime!!

    val milliseconds: Double
        get() = time.toDouble() / 1000000.0

    fun reset(): TimingLogger
    {
        startingTime = System.nanoTime()
        return this
    }

    fun init(): TimingLogger
    {
        return reset()
    }

    fun clear()
    {
        startingTime = null
    }
}
